core/backend/src/test/standalone/TxnManager.test.ts

```diff
@@ -920,3 +945,116 @@ describe("TxnManager", () => {
+
+      assertTxnStates(txns[0], true, true, false, false, false, "initial create");
+      briefcases[0].saveChanges("inserted first model");
+      assertTxnStates(txns[0], true, false, true, true, false, "initial saved changes");
+
+      await briefcases[0].pushChanges({ accessToken: accessTokens[0], description: "new physical model" });
+      assertTxnStates(txns[0], false, false, false, false, false, "pushed changes");
+
+      // [1] model has no changes
+      await briefcases[1].pullChanges();
+      assertTxnStates(txns[1], false, false, false, false, false, "pulled changes");
+
+      // [2] model has saved changes
+      IModelTestUtils.createAndInsertPhysicalPartitionAndModel(briefcases[2], IModelTestUtils.getUniqueModelCode(briefcases[2], "localPhysicalModel"), false);
+      briefcases[2].saveChanges("inserted local model");
+      assertTxnStates(txns[2], true, false, true, true, false, "briefcase 2 with saved local changes");
+      assert.strictEqual(txns[2].getUndoString(), "inserted local model");
+
+      await briefcases[2].pullChanges();
+      if(pullMergeMethod === "Merge") {
+        assert.strictEqual(txns[2].getUndoString(), "Merged");
+        assertTxnStates(txns[2], true, false, true, true, false, "briefcase 2 after pull");
+      } else {
+        assertTxnStates(txns[2], true, false, true, false, false, "briefcase 2 after pull");
+      }
+      assert.strictEqual(2, getCount(briefcases[2], PhysicalModel.classFullName)); // the pulled and local models
+
+      // [3] model gets unsaved changes
+      IModelTestUtils.createAndInsertPhysicalPartitionAndModel(briefcases[3], IModelTestUtils.getUniqueModelCode(briefcases[3], "localPhysicalModel"), false);
+      assertTxnStates(txns[3], true, true, false, false, false, "briefcase 3 with unsaved local changes");
+      await assertThrowsAsync(async () => briefcases[3].pullChanges(), "cannot processed due to unsaved changes.");
+      assertTxnStates(txns[3], true, true, false, false, false, "briefcase 3 after pull with unsaved changes");
+
+      // [3] model: now save, undo, pull and redo
+      briefcases[3].saveChanges("Temporarily save changes");
+      assertTxnStates(txns[3], true, false, true, true, false, "briefcase 3 with saved local changes");
+      assert.strictEqual(1, getCount(briefcases[3], PhysicalModel.classFullName)); // the local model
+      assert.strictEqual(txns[3].reverseTxns(1), IModelStatus.Success);
+      assertTxnStates(txns[3], false, false, false, false, true, "briefcase 3 after undo");
+      assert.strictEqual(0, getCount(briefcases[3], PhysicalModel.classFullName)); // no model
+      await briefcases[3].pullChanges();
+
+      assert.strictEqual(1, getCount(briefcases[3], PhysicalModel.classFullName));
+      assertTxnStates(txns[3], false, false, false, false, true, "briefcase 3 after pull");
+      // TODO: "redo possible" should be false, if we try to reinstate, it will crash!
```

Comment:
This is currently a bug it seems. "IsRedoPossible" should be false (at least for merge.
Calling ReinstateTxn() after this crashes node.